* Библиотека requests

В современном веб-программировании взаимодействие между клиентом и сервером на основе протокола HTTP является одной из ключевых задач. Для разработчиков на языке Python библиотека requests стала стандартом де-факто для выполнения HTTP-запросов благодаря своей простоте и удобству. В этой статье мы рассмотрим, что такое библиотека requests, а также охватим основы работы с HTTP/S.

Что такое библиотека requests?

requests — это мощная и популярная библиотека для языка Python, разработанная для удобного выполнения HTTP-запросов. Она позволяет вам взаимодействовать с веб-ресурсами, отправлять запросы и обрабатывать ответы, используя простые и лаконичные методы. Библиотека была создана, чтобы сделать процесс работы с HTTP более интуитивно понятным, чем традиционные методы, такие как стандартный модуль urllib.

Некоторые ключевые особенности библиотеки requests включают:

Легкость в использовании: библиотека имеет простой и понятный интерфейс.
Поддержка HTTP-методов: GET, POST, PUT, DELETE, и многие другие.
Работа с JSON: простое преобразование данных между Python-объектами и JSON.
Обработка сессий: поддержка управления сессиями и куки через объект Session.
Управление заголовками и параметрами запроса: возможность легко добавлять параметры и заголовки к запросам.
Работа с сертификатами: поддержка HTTPS и безопасных соединений.Основы работы с HTTP


Что такое HTTP?
HTTP (HyperText Transfer Protocol) — это протокол, который используется при обмене данными в интернете. Он основан на модели клиент-сервер, где клиент (обычно веб-браузер) отправляет запросы к серверу, а сервер отвечает на эти запросы. ПротоколHTTP имеет множество методов, среди которых наибольшее распространение получили:

GET: используется для получения данных с сервера. Например, запрос на получение веб-страницы.
POST: применяется для отправки данных на сервер, например, при отправке формы на веб-сайте.
PUT: предназначен для обновления существующих ресурсов на сервере.
DELETE: используется для удаления ресурсов с сервера.
Когда клиент выполняет HTTP-запрос, он отправляет запрос, который может содержать заголовки, параметры и тело (особенно для методов POST и PUT). Сервер обрабатывает запрос и отправляет обратно ответ, который включает статусный код, заголовки и, возможно, тело ответа (например, HTML-код страницы или данные в формате JSON).

Выполнение запросов с библиотекой requests
Теперь давайте рассмотрим, как использовать библиотеку requests для выполнения различных типов запросов.

Установка библиотеки
Чтобы установить библиотеку requests, вы можете использовать pip:

In [ ]:
# pip install requests

Простой GET-запрос
Выполнение GET-запроса очень просто. Например, чтобы получить данные с определенного URL, вы можете сделать следующее:

In [ ]:
import niquests as requests

response = requests.get('https://yandex.ru')
print(f'Status code: {response.status_code}') # Выводит статусный код ответа
print()
print(f'Text: {response.text}') # Выводит текст ответа

Status code: 200

Text: <body></body><script nonce='15c9353b76115ff0d1d5f16cd49eb075'>var it = {"host":"https:\u002F\u002Fsso.dzen.ru\u002Finstall?uuid=55f52ad7-cc8f-44c5-a536-6f4c10424bf8","retpath":"https:\u002F\u002Fdzen.ru\u002F?yredirect=true&is_autologin_ya=true"};(function() { var form = document.createElement('form'); var element1 = document.createElement('input'); var element2 = document.createElement('input'); element1.name = 'retpath'; element1.type = 'hidden'; element1.value = it.retpath; form.appendChild(element1); element2.name = 'container'; element2.type = 'hidden'; element2.value = '1744828542.10154097.ihZk_wpBtKi5-Ioh.YlQvFJeC-q2HYo7_yHeeM1qhPqdNVuzTsIVG_Wp3WknkbTaw3618PlyWCvO2lKQ3dv13jfVOQ71NNPSNBoafn4KzrpsGNm7MnTekiro0h-4M16Lu45eASwESXv3QLY-Neu1Y-A7X47ORQSWsyhlruvdSmH06ekvOUcOlmyWXs7CoRAfi8xHsLwXGDwJYCptcl4RIgoY-Bimn-E7W3ybKDENSSqBE8ZLUP6JnCdXuTZd5OzH8kwHu_--UNzrWfpEu1HTB51c2g5SQkS6sXWqjs6EqfTvoqsRsNnjjTqgpzL3ME6gDBa-zNikYTSFkRv8ZqF5IM4sTqQ1zP_rUO3LM-qTgjgtM1OYfmOsV

POST-запрос

In [2]:
import requests

data = {'key' : 'value'}
response = requests.post('https://api.example.com/submit', json=data)
print(response.status_code)
print(response.json()) # Если ответ в формате JSON

ConnectionError: HTTPSConnectionPool(host='api.example.com', port=443): Max retries exceeded with url: /submit (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x0000020300024190>: Failed to resolve 'api.example.com' ([Errno 11001] getaddrinfo failed)"))

Обработка ошибок
При работе с HTTP-запросами важно обрабатывать возможные ошибки. Библиотека requests предоставляет удобные методы для этого:

In [3]:
import requests

try:
  response = requests.get ('https://api. example.com/data' )
  response.raise_for_status() # Вызывает исключение для ошибок НТТР
except requests. exceptions. HTTPError as err:
  print (f"HTTP error occurred: {err}") # 06pa60TKa own60K
except Exception as err:
  print (f"An error occurred: {err}")

An error occurred: HTTPSConnectionPool(host='api.%20example.com', port=443): Max retries exceeded with url: /data (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x0000020300F38410>: Failed to resolve 'api.%20example.com' ([Errno 11001] getaddrinfo failed)"))


1. Отправка GET-запросов

GET — это основной HTTP-метод, который используется для запроса данных с сервера. Обычно запросы GET используются для получения информации без изменения состояния ресурса на сервере.

Для отправки GET-запроса в библиотеке requests достаточно вызвать функцию requests.get(). Например:

In [4]:
import os
import requests

response = requests.get ('https: //jsonplaceholder.typicode.com/posts' )
print(response.status_code) # Выводим код состояния ответа
print(response.text)   # Выводим текст ответа
print(response.json()) # Выводим ответ в формате JSON

InvalidURL: Invalid URL 'https: //jsonplaceholder.typicode.com/posts': No host supplied

Параметры запроса:

In [5]:
import requests

params = {'userId': 1}
response = requests.get('https://jsonplaceholder.typicode.com/posts', params=params)
print(response.json())

[{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}, {'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}, {'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}, {'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident rerum culpa\nquis hic c

2. Отправка POST-запросов

POST-запросы часто используются для отправки данных на сервер, например, для создания новых записей в базе данных. В библиотеке requests для отправки POST-запроса используется метод requests.post().

Пример отправки POST-запроса:

In [6]:
data = {
  'title': 'foo',
  'body': 'bar',
  'userId': 1
}

response = requests.post('https://jsonplaceholder.typicode.com/posts', data=data)

# Выводим код состояния ответа
print(response.status_code)

# Выводим данные, отправленные сервером в ответ
print(response.json())

201
{'title': 'foo', 'body': 'bar', 'userId': '1', 'id': 101}


Параметры запроса:

Для отправки данных, например, в формате JSON, можно использовать параметр json:

In [7]:
import requests

data = {
  'title': 'foo',
  'body' : 'bar',
  'userId': 1
}

response = requests.post('https://jsonplaceholder.typicode.com/posts', json=data)

print(response.json())

{'title': 'foo', 'body': 'bar', 'userId': 1, 'id': 101}


3. Другие HTTP-методы: PUT, PATCH, DELETE

В библиотеке requests также поддерживаются другие HTTP-методы, такие как PUT, PATCH и DELETE.

PUT-запросы

PUT используется для обновления существующего ресурса на сервере. Он заменяет весь ресурс новыми данными.

Пример отправки PUT-запроса:

In [8]:
import requests

data = {
  'id': 1,
  'title': 'Updated Title',
  'body': 'Updated Body' ,
  'userId': 1
}

response = requests.put('https://jsonplaceholder.typicode.com/posts/1', json=data)

print(response.status_code)
print(response.json())

200
{'id': 1, 'title': 'Updated Title', 'body': 'Updated Body', 'userId': 1}


PATCH-запросы

PATCH — это метод, который используется для частичного обновления ресурса. В отличие от PUT, который заменяет весь ресурс, PATCH позволяет обновить только некоторые поля.

Пример отправки PATCH-запроса:

In [9]:
import requests

data = {
  'title': 'Updated Title'
}

response = requests.patch('https://jsonplaceholder.typicode.com/posts/1', json=data)

print(response.status_code)
print(response.json())

200
{'userId': 1, 'id': 1, 'title': 'Updated Title', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}


DELETE-запросы

DELETE используется для удаления ресурса на сервере. Этот метод может быть использован, чтобы удалить данные по определённому URL.

Пример отправки DELETE-запроса:

In [10]:
import requests

response = requests.delete('https://jsonplaceholder.typicode.com/posts/1')

print(response.status_code) # Ожидаемый код состояния: 200 (ОК), если удаление прошло успешно

200


Работа с ответами в библиотеке requests в Python

Работа с ответами является важной частью взаимодействия с веб-сервисами, и библиотека requests предоставляет удобные средства для обработки ответов от серверов. В этой статье мы рассмотрим, как обрабатывать ответы, проверять успешность запросов и обрабатывать возможные ошибки.

1. Обработка ответов

После отправки HTTP-запроса с помощью библиотеки requests, сервер отправляет ответ, который доступен через объект Response. Этот объект содержит информацию о статусе запроса, теле ответа и другие полезные данные.

Пример отправки запроса и получения ответа:

In [11]:
import requests

response = requests.get('https://jsonplaceholder.typicode.com/posts')

# Получение кода состояния ответа
print("Status Code:", response.status_code)

# Получение текста ответа
print("Response Text:", response.text)

# Получение данных в формате JSON (если они есть)
print("Response JSON:", response.json())

Status Code: 200
Response Text: [
  {
    "userId": 1,
    "id": 1,
    "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
    "body": "quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto"
  },
  {
    "userId": 1,
    "id": 2,
    "title": "qui est esse",
    "body": "est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla"
  },
  {
    "userId": 1,
    "id": 3,
    "title": "ea molestias quasi exercitationem repellat qui ipsa sit aut",
    "body": "et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut"
  },
  {
    "userId": 1,
    "id": 4,
    "title": "eum et est occaecati",


Что можно извлечь из объекта response?

status_code: Код состояния HTTP-ответа, например, 200 для успешного запроса.
text: Тело ответа в виде строки (может быть HTML, JSON и т.д.).
json(): Преобразует тело ответа в формат JSON (если это возможно).
headers: Заголовки ответа.
cookies: Куки, если они присутствуют в ответе.
url: URL-адрес, по которому был отправлен запрос.
Пример вывода:

In [ ]:
Status Code: 200
Response Text: [{"userId": 1, "id": 1, "title": "Post 1", ... }]
Response JSON: [{"userId": 1, "id": 1, "title": "Post 1", ... }]


2. Проверка успешности запроса

Для проверки, был ли запрос успешным, можно использовать код состояния ответа, который возвращается сервером. Обычно, если код состояния начинается с 2xx, запрос прошел успешно.

200 OK — успешный запрос.
201 Created — ресурс был успешно создан.
400 Bad Request — ошибка запроса на сервере.
404 Not Found — ресурс не найден.
В библиотеке requests есть несколько удобных методов, чтобы проверить успешность запроса:

Проверка через status_code:

In [12]:
import requests

response = requests.get('https://jsonplaceholder.typicode.com/posts')

if response.status_code == 200:
  print ("Request was successful!")
else:
  print (f"Request failed with status code {response. status_code}")

Request was successful!


Использование response.ok:

Библиотека requests предоставляет атрибут ok, который возвращает True, если код состояния ответа указывает на успешный запрос (коды от 200 до 299):

In [13]:
import requests

response = requests.get('https://jsonplaceholder.typicode.com/posts')

if response.ok:
  print ("Request was successful!")
else:
  print (f"Request failed with status code {response. status_code}")

Request was successful!


3. Обработка ошибок

Когда запрос не может быть выполнен (например, сервер не доступен), будет выброшено исключение. Для обработки этих исключений можно использовать блоки try/except.

Пример обработки ошибок:

In [15]:
import requests
from requests.exceptions import HTTPError, Timeout, RequestException

try:

  response.raise_for_status() # Проверка на успешный код ответа (напри
except HTTPError as http_err:
  print (f"HTTP error occurred: {http_err}")
except Timeout:
  print ("The request timed out")
except RequestException as err:
  print (f"Other error occurred: {err}")
else:
  print ("Request was successful!")
  print(response.json() )


Request was successful!
[{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}, {'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}, {'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}, {'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident

HTTPError: Ошибка, связанная с кодом ответа сервера (например, 404 или 500).
Timeout: Ошибка, если запрос занял слишком много времени.
RequestException: Общая ошибка, которая возникает при проблемах с сетевым соединением, например, если сервер недоступен.

4. Проверка наличия данных в ответе

Иногда вам нужно убедиться, что сервер вернул данные в ответе, и корректно обработать отсутствие этих данных. Например, если сервер возвращает пустой список или ошибку, вы можете обработать это соответствующим образом.

In [16]:
import requests

response = requests.get('https://jsonplaceholder.typicode.com/posts')

if response.ok:
  data = response. json()
  if not data:
    print("No data found")
  else:
    print ("Data received:", data)
else:
  print(f"Request failed with status code {response. status_code}" )


Data received: [{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}, {'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}, {'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}, {'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident rerum cu

Заголовки и куки

Заголовки и куки — это важные части HTTP-запросов, которые часто используются для передачи метаинформации, таких как тип содержимого, информация о браузере, авторизация и сессии.

Работа с заголовками

Заголовки позволяют клиенту сообщать серверу важную информацию. Например, мы можем указать, что отправляем JSON, или передать информацию о пользовательском агенте.

Пример передачи заголовков в запросе:

In [17]:
import requests

# Определение заголовков
headers = {
  'User-Agent': 'my-app/1.0',
  'Accept': 'application/json',
  'Authorization': 'Bearer your_token'
}

response = requests.get('https://jsonplaceholder.typicode.com/posts', headers=headers)

print(response.status_code)
print(response.text)

200
[
  {
    "userId": 1,
    "id": 1,
    "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
    "body": "quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto"
  },
  {
    "userId": 1,
    "id": 2,
    "title": "qui est esse",
    "body": "est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla"
  },
  {
    "userId": 1,
    "id": 3,
    "title": "ea molestias quasi exercitationem repellat qui ipsa sit aut",
    "body": "et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut"
  },
  {
    "userId": 1,
    "id": 4,
    "title": "eum et est occaecati",
    "body": "ullam et saepe 

В этом примере мы передаем заголовки, которые указывают серверу, что клиент — это приложение с именем my-app/1.0, и что мы ожидаем ответ в формате JSON. Также передаем токен авторизации через заголовок Authorization.

Работа с куками

Куки используются для хранения информации о состоянии между запросами, например, для сохранения сессии пользователя.

Пример работы с куками:

In [18]:
import requests

# Создание куки
cookies = {'session_id': '123456789' }

# Отправка запроса с куками
response = requests.get ('https://jsonplaceholder.typicode.com/posts', cookies=cookies)

print(response.status_code)
print (response.cookies)

200
<RequestsCookieJar[]>


В этом примере мы отправляем запрос с кукой session_id, которая может быть использована сервером для отслеживания состояния сессии пользователя.

2. Передача параметров и заголовков в запросах

Параметры часто используются в URL-адресах для уточнения запроса (например, фильтры или лимиты). Библиотека requests позволяет легко передавать параметры и заголовки в запросах.

Передача параметров в запросах
Для передачи параметров в запросах GET, POST и других, библиотека requests позволяет использовать параметр params. Этот параметр принимает словарь, который автоматически преобразуется в строку запроса.

Пример:

In [19]:
import requests

# Параметры запроса
params = {
'userId': 1,
'limit': 5

}

response = requests.get ('https://jsonplaceholder.typicode.com/posts' , params=params)

print(response.status_code)
print (response. json ( ) )

200
[{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}, {'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}, {'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}, {'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident rerum culpa\nquis h

Здесь мы передаем параметры userId и limit, которые будут автоматически добавлены к URL как часть строки запроса: ?userId=1&limit=5.

Передача параметров в запросе POST
В случае с методом POST, параметры передаются в теле запроса:


In [20]:
import requests

# Данные для отправки
data = {
'title': 'New Post',
'body': 'This is the body of the post',
'userId': 1

}

response = requests.post ('https://jsonplaceholder.typicode.com/posts', data=data)

print(response.status_code)
print (response. json () )

201
{'title': 'New Post', 'body': 'This is the body of the post', 'userId': '1', 'id': 101}


Здесь мы отправляем данные в теле запроса, и сервер обрабатывает их как форму.

3. Сессии и авторизация
Сессии и авторизация необходимы для работы с веб-сервисами, которые требуют подтверждения личности пользователя или сохранения состояния между запросами.

Использование сессий
Сессия позволяет сохранять параметры, такие как куки, авторизационные данные и заголовки, между несколькими запросами. Это полезно, когда нужно работать с одним и тем же сервером в рамках одного сеанса.

Пример:

In [22]:
import requests

# Создание сессии
session = requests. Session ( )

# Добавление заголовков в сессию
session.headers. update({'User-Agent': 'my-app/1.0'})

# Отправка первого запроса
response = session.get('https: //jsonplaceholder.typicode. com/posts' )
print(response.status_code)

# Отправка второго запроса с той же сессией
response = session. get('https: //jsonplaceholder. typicode. com/posts/1')
print (response.status_code)

InvalidURL: Invalid URL 'https: //jsonplaceholder.typicode. com/posts': No host supplied

В этом примере мы создаем сессию, добавляем заголовки и отправляем несколько запросов в рамках одной сессии. Это позволяет эффективно управлять состоянием.

Авторизация
Авторизация необходима для получения доступа к защищенным ресурсам. Библиотека requests позволяет использовать базовую авторизацию и токены.

Базовая авторизация


In [23]:
import requests
from requests.auth import HTTPBasicAuth

response = requests.get('https: //jsonplaceholder.typicode.com/posts', auth=HTTPBasicAuth('username', 'password' ) )

print(response.status_code)
print (response. text)

InvalidURL: Invalid URL 'https: //jsonplaceholder.typicode.com/posts': No host supplied

Авторизация с использованием токена

In [24]:
import requests

headers = {
'Authorization': 'Bearer your_token'

}

response = requests.get ('https://jsonplaceholder.typicode.com/posts' , headers=headers)

print(response.status_code)
print (response. text)

200
[
  {
    "userId": 1,
    "id": 1,
    "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
    "body": "quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto"
  },
  {
    "userId": 1,
    "id": 2,
    "title": "qui est esse",
    "body": "est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla"
  },
  {
    "userId": 1,
    "id": 3,
    "title": "ea molestias quasi exercitationem repellat qui ipsa sit aut",
    "body": "et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut"
  },
  {
    "userId": 1,
    "id": 4,
    "title": "eum et est occaecati",
    "body": "ullam et saepe 

В этом примере используется авторизация с помощью токена Bearer, который добавляется в заголовок запроса.